### ISSUE TITLE
--auto-gen-config can fail with "Error: empty section found in .rubocop_todo.yml"

### ISSUE THREAD

**Issue by dominicsayers** [bbatsov/rubocop#5281]
When running Rubocop with the `--auto-gen-config` option, the part-built output file `.rubocop_todo.yml` is loaded and checked during the processing of each cop (because the generated excludes are added to any defined manually in `.rubocop.yml` and when `rubocop.yml` is loaded it follows the inheritance down to `.rubocop_todo.yml`).

One [validation check is for an empty section](https://github.com/bbatsov/rubocop/blob/v0.52.0/lib/rubocop/config.rb#L426).

The output file is built up cop-by-cop during this process by writing to it after each cop using [`puts`](https://github.com/bbatsov/rubocop/blob/v0.52.0/lib/rubocop/formatter/disabled_config_formatter.rb#L79). Ruby's caching of these writes means that sometimes only a partial image of the offences is written to the file and this partial image is what is read during the next check.

Occasionally, by coincidence, the end of the buffer is the beginning of a cop section like this:

``` yaml
Style/CaseEquality:
  Exclude:
```

and the empty `Exclude` section causes the validation error

```
Error: empty section Style/CaseEquality found in .rubocop_todo.yml
```

### Possible solutions

1.  Add an `output.flush` after each `output.puts`. This works and is easy to implement but is maybe slows things down a bit.
1.  Do not write `.rubocop_todo.yml` while it's being built. Instead write to a temporary file and rename it after building has finished.
1.  Add a config load option that ignores the part-built todo file.

--------

## Expected behavior

I expect `rubocop --auto-gen-config` to complete successfully.

## Actual behavior

Instead, `rubocop --auto-gen-config` exits with an error status and is truncated at the end of the output buffer last written.

## Steps to reproduce the problem

It's going to be difficult to reproduce this because it depends on the flushing behaviour of Ruby's `IO` class. It happened to me with my repository which is large and contains many offences, but it's private unfortunately.

I'm happy to help reproduce this any way I can

## RuboCop version

```
$ rubocop -V
0.52.0 (using Parser 2.4.0.2, running on ruby 2.2.2 x86_64-darwin16)
```

**Comment by bquorning** [comment]
I experienced the same issue yesterday, while running rubocop for the first time on a large project, with many offenses. And I, too, suspected a puts/flush issue, but didn’t investigate as thoroughly as @dominicsayers.

Reverting back to v0.51.0 allowed me to run `rubocop --auto-gen-config`, so the problem is probably somewhere between the last two releases.

**Comment by bquorning** [comment]
See #5294 for a potential solution.

**Comment by dominicsayers** [comment]
Checking now with my code

**Comment by dominicsayers** [comment]
OK, here's what happened using the #5294 code with my codebase.

1.  `--auto-gen-config` completed without error
1.  The generated file `.rubocop_todo.yml` usually has its keys in alphabetical order. In the one that was generated by #5294, the key `Metrics/LineLength` appears at the end, out of order. I was not expecting to see this key at all because this cop is disabled in my `.rubocop.yml`.
1.  The list of exclusions for `Style/IfUnlessModifier` was incomplete so when I ran `rubocop` after generating the to-do file there were still some offences reported for this cop.

I can't say whether these two problems were caused by your changes or other issues with Rubocop post-0.52 - my guess would be it's other people's changes not yours. I can't check because the master branch won't complete successfully against my codebase. 

cc @bquorning

**Comment by bquorning** [comment]
1. Phew 😄 
2. Looking at the spec changes in 9e76387952725f1a7fa54989024114549c0b7d80, I think `Metrics/LineLength` is now expected to be at the end of the auto-generated file.
3. I’ve seen the same issue, only for `Naming/MethodName`. I think these are separate issues though.
